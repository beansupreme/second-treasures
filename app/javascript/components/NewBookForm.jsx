import React from "react";
import axios from "axios";
import ErrorList from "./ErrorList"
import MessageBox from "./MessageBox"

const defaultState = {
  title: '',
  author: '',
  price: 0.00,
  isbn: '',
  genre: '',
  errors: [],
  message: ''
}

class NewBookForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = defaultState

    this.handleSubmit = this.handleSubmit.bind(this);
    this.updateField = this.updateField.bind(this);
  }

  handleSubmit(event) {
    event.preventDefault();
    axios.post(
      '/api/v1/books',
      {
        book: this.buildBook()        
      }
    ).then(response => {
      this.props.onNewBookAdd(response.data);
      
      this.setState(defaultState);
      this.setState({message: `'${response.data.title}' successfully added`});
      setTimeout(() => {
        this.setState({message: ''})
      }, 5000);

    }).catch(error => {
      let response = error.response;
      if (response.status === 422) {
        this.setState({
          errors: response.data,
          message: ''
        });
      } else {
        this.setState({
          errors: ['Something went wrong... Books cannot be created at this time.'],
          message: ''
        });
      }
    })
  }

  buildBook() {
    return {
      title: this.state.title,
      author: this.state.author,
      price: this.state.price,
      isbn: this.state.isbn,
      genre: this.state.genre
    }
  }


  updateField(event) {
    let fieldName = event.target.id.split('_')[1];
    let fieldsToUpdate = {};
    fieldsToUpdate[fieldName] = event.target.value;
    this.setState(fieldsToUpdate);
  }

  render() {
    return (
      <React.Fragment>
        <h5 className="text-center">Add a new Book recommendation:</h5>
        <ErrorList id="new-book-form-errors" errors={this.state.errors} />
        <MessageBox id="new-book-form-message" message={this.state.message} />
        <form className="form-inline" id="new-book-form" onSubmit={this.handleSubmit}>
          <div className="form-group">
            <input id="book_title_field" name="title" type="text" placeholder="Title" className="form-control" 
            onChange={this.updateField} value={this.state.title}/>
          </div>
          <div className="form-group">
            <input id="book_author_field" name="author" type="text" placeholder="Author" className="form-control" 
            onChange={this.updateField} value={this.state.author}/>
          </div>
          <div className="form-group">
            <input id="book_price_field" name="price" type="number" placeholder="Price" className="form-control" 
             min="0" max="10000" step="0.01" onChange={this.updateField} value={this.state.price}/>
          </div>
          <div className="form-group">
            <input id="book_isbn_field" name="ISBN" type="text" placeholder="ISBN" className="form-control" 
            onChange={this.updateField} value={this.state.isbn}/>
          </div>
          <div className="form-group">
            <input id="book_genre_field" name="Genre" type="text" placeholder="Genre" className="form-control" 
            onChange={this.updateField} value={this.state.genre}/>
          </div>
          <button type="submit" className="btn btn-success">Save</button>
        </form>  
      </React.Fragment>
    )
  }
}

NewBookForm.defaultProps = {
  onNewBookAdd: () => {}
}

export default NewBookForm;