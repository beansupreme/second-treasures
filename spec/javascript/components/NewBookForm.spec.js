import React from 'react';
import renderer from 'react-test-renderer';
import { mount, shallow} from 'enzyme';
import sinon from 'sinon';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import NewBookForm from 'components/NewBookForm';

describe('<NewBookForm />', () => { 
  var mock = new MockAdapter(axios);
  const newBookResponse = {
    id: 14,
    title: 'By Gaslight',
    author: 'Steven Price',
    price: 12.32,
    isbn: '978-0060850666'
  }
  beforeEach(() =>{

    mock.onPost('/api/v1/books').reply(200, 
      newBookResponse
    );
  });

  it('renders as expected', () => {
    const tree = renderer.create(<NewBookForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });


  it('calls to the books create endpoint with the current state', () => {
    sinon.spy(axios, 'post');

    const wrapper = mount(<NewBookForm />);

    wrapper.setState({
      title: 'By Gaslight',
      author: 'Steven Price',
      price: 12.32,
      isbn: '978-0060850666'
    });

    wrapper.find('#new-book-form').first().simulate('submit');

    expect(axios.post.calledOnce).toEqual(true);

    let expectedPostBody = {
      book: {
        title: 'By Gaslight',
        author: 'Steven Price',
        price: 12.32,
        isbn: '978-0060850666'
      } 
    };

    let postData = axios.post.getCall(0).args;
    
    expect(postData[0]).toEqual('/api/v1/books');
    expect(postData[1]).toEqual(expectedPostBody);

    expect(axios.post.calledOnce).toEqual(true);

    let getData = axios.post.getCall(0).args;


    axios.post.restore();
  });

  it('updates state when the inputs are changed', () => {
    const wrapper = mount(<NewBookForm />);

    updateInput(wrapper, 'book_title_field', 'The Game');
    updateInput(wrapper, 'book_author_field', 'Ken Dryden');
    updateInput(wrapper, 'book_price_field', '22.78');
    updateInput(wrapper, 'book_isbn_field', '978-29292929');


    let state = wrapper.state();
    const expectedState = {
      title: 'The Game',
      author: 'Ken Dryden',
      price: '22.78',
      isbn: '978-29292929',
      errors: [],
      message: ''
    }
    expect(state).toEqual(expectedState)
  });

  it('resets state after a book is created', (done) => {
    const wrapper = mount(<NewBookForm />);

    wrapper.setState({
      title: 'By Gaslight',
      author: 'Steven Price',
      price: 12.32,
      isbn: '978-0060850666'
    });

    wrapper.find('#new-book-form').first().simulate('submit');
    

    setTimeout(() => {
      expect(wrapper.state()).toEqual({
        title: '',
        author: '',
        price: 0,
        isbn: '',
        errors: [],
        message: "'By Gaslight' successfully added"
      });
      done()
    }, 0);
  });

  it('displays a success message if create book succeeds', (done) => {
    const wrapper = mount(<NewBookForm />);

    wrapper.find('#new-book-form').first().simulate('submit');


    setTimeout(() => {
      expect(wrapper.find('#new-book-form-message').first()).toIncludeText("'By Gaslight' successfully added");
      expect(wrapper.find('#new-book-form-errors')).toHaveHTML('<div></div>');
      done()
    }, 0);

    
  });

  it('renders errors on the page if the create book call fails', (done) => {
    const wrapper = mount(<NewBookForm />);

    mock.onPost('/api/v1/books').reply(422, 
      ["Title can't be blank"]
    );

    wrapper.find('#new-book-form').first().simulate('submit');

    setTimeout(() => {
      expect(wrapper.find('#new-book-form-message').first()).toHaveHTML('<div></div>');
      expect(wrapper.find('#new-book-form-errors').first()).toIncludeText("Title can't be blank");
      done();
    }, 0);

    mock.restore();
  });

  xit('calls the onNewBookAdd callback when a new book is submitted', (done) => {
    let handleNewBookAdd = sinon.spy();

    const wrapper = mount(<NewBookForm onNewBookAdd={handleNewBookAdd}/>);

    wrapper.setState({
      title: 'By Gaslight',
      author: 'Steven Price',
      price: 12.32,
      isbn: '978-0060850666'
    });

    wrapper.find('#new-book-form').first().simulate('submit');

    setTimeout(() => {
      expect(handleNewBookAdd.called).toEqual(true)

      expect(handleNewBookAdd.calledWith({title: 'By Gaslight'})).toEqual(true)
      done();
    }, 0);
  });

});

function updateInput(wrapper, inputId, value) {
  let input = wrapper.find('#' + inputId).first();
  input.simulate('change', { target: { id: inputId, value: value }});
}