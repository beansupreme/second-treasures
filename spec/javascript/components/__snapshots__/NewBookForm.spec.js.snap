// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewBookForm /> renders as expected 1`] = `
Array [
  <h5
    className="text-center"
  >
    Add a new Book recommendation:
  </h5>,
  <div />,
  <div>
    
  </div>,
  <form
    className="form-inline"
    id="new-book-form"
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        className="form-control"
        id="book_title_field"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        id="book_author_field"
        name="author"
        onChange={[Function]}
        placeholder="Author"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        id="book_price_field"
        max="10000"
        min="0"
        name="price"
        onChange={[Function]}
        placeholder="Price"
        step="0.01"
        type="number"
        value={0}
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        id="book_isbn_field"
        name="ISBN"
        onChange={[Function]}
        placeholder="ISBN"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        id="book_genre_field"
        name="Genre"
        onChange={[Function]}
        placeholder="Genre"
        type="text"
        value=""
      />
    </div>
    <button
      className="btn btn-success"
      type="submit"
    >
      Save
    </button>
  </form>,
]
`;
